tritium-proxy/src/main/java/com/palantir/tritium/proxy/Proxies.java

```diff
@@ -59,27 +60,27 @@ static Class<?>[] interfaces(Class<?> iface, Class<?> delegateClass) {
         checkIsInterface(iface);
         Objects.requireNonNull(delegateClass, "delegateClass");
 
-        Set<Class<?>> interfaces = new LinkedHashSet<>();
-        interfaces.add(iface);
         if (delegateClass.isInterface()) {
-            interfaces.add(delegateClass);
-        } else {
-            interfaces.addAll(Arrays.asList(delegateClass.getInterfaces()));
+            if (iface.equals(delegateClass)) {
+                return new Class<?>[] {iface};
+            }
+            return new Class<?>[] {iface, delegateClass};
         }
 
-        checkAreAllInterfaces(interfaces);
-        return interfaces.toArray(new Class<?>[0]);
+        Class<?>[] delegateInterfaces = delegateClass.getInterfaces();
+        int expectedSize = delegateInterfaces.length + 1;
+        Set<Class<?>> interfaces = new LinkedHashSet<>((int) ((float) expectedSize / 0.75F + 1.0F));
+        interfaces.add(iface);
+        for (Class<?> possibleInterface : delegateInterfaces) {
+            interfaces.add(checkIsInterface(possibleInterface));
```

Comment:
`possibleInterface` may be `iface` (or a super/sub-type of an interface we've already added). Allowed by the contract, depending what we're optimizing for it may or may not be worth changing.

Reply:
Are you suggesting we should coalesce to the minimum set
Of interfaces? Since this is a set, we won't get duplicates, but we effectively union all of the interface types & super types.

Reply:
Just a stream of consciousness — in the bytebuddy implementation I think we try to find the minimal set (after the primary interface, anyhow), but that’s probably not necessary here